### PR TITLE
[SPARK-38579][SQL][WEBUI]Requesting Restful API can cause NullPointerException

### PR DESCRIPTION
**Pull request by yimin-yang**
### What changes were proposed in this pull request?

Added null check for `exec.metricValues`.

### Why are the changes needed?

When requesting Restful API  {baseURL}/api/v1/applications/$appId/sql/$executionId which is introduced by this PR https://github.com/apache/spark/pull/28208, it can cause NullPointerException. The root cause is, when calling method doUpdate() of `LiveExecutionData`, `metricsValues` can be null. Then, when statement `printableMetrics(graph.allNodes, exec.metricValues)` is executed, it will throw NullPointerException.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Tested manually.

### CODE REVIEW

**Review by weixiuli — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala [hunk 1]
```diff
@@ -94,7 +95,15 @@ private[v1] class SqlResource extends BaseAppResource {
 
     val duration = exec.completionTime.getOrElse(new Date()).getTime - exec.submissionTime
     val planDetails = if (planDescription) exec.physicalPlanDescription else ""
-    val nodes = if (details) printableMetrics(graph.allNodes, exec.metricValues) else Seq.empty
```
Comment (weixiuli) [inline sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala:97]: Maybe we can make a small change，like:

``` val nodes = if (details) printableMetrics(graph.allNodes,  Option(exec.metricValues).getOrElse(new HashMap[Long, String])) else Seq.empty```

File: sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala [hunk 2]
```diff
@@ -94,7 +95,15 @@ private[v1] class SqlResource extends BaseAppResource {
...
+    val nodes = if (details) {
+      if (exec.metricValues != null) {
+        printableMetrics(graph.allNodes, exec.metricValues)
+      } else {
+        printableMetrics(graph.allNodes, new HashMap[Long, String])
+      }
```
Comment (mridulm) [inline sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala:103]: `Option(exec.metricValues).getOrElse(Map())` instead

**Comment by yym1995** [comment]
@mridulm @weixiuli Please review this again, thanks!

File: sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala [hunk 3]
```diff
@@ -94,7 +94,8 @@ private[v1] class SqlResource extends BaseAppResource {
 
     val duration = exec.completionTime.getOrElse(new Date()).getTime - exec.submissionTime
     val planDetails = if (planDescription) exec.physicalPlanDescription else ""
-    val nodes = if (details) printableMetrics(graph.allNodes, exec.metricValues) else Seq.empty
+    val nodes = if (details) printableMetrics(graph.allNodes,
+      Option(exec.metricValues).getOrElse(Map())) else Seq.empty
```
Comment (martin-g) [inline sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala:98]:
```suggestion
      Option(exec.metricValues).getOrElse(Map.empty)) else Seq.empty
```

**Review by martin-g — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala (hunk 3 above)
Comment (martin-g) [reply sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala]: nit: I wouldn't use `Option` here. Just check for `null`, e.g. `val metricValues = if (exec.metricValues != null) exec.metricValues else Map.empty`
But I guess the extra memory allocation for `Option` is not a big deal here

**Comment by yym1995** [comment]
@martin-g Could you merge this PR? Thx

**Comment by martin-g** [comment]
I am just a contributor here.
Someone from the Spark team will merge it soon!

File: sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala [hunk 4]
```diff
@@ -94,7 +94,8 @@ private[v1] class SqlResource extends BaseAppResource {
 
     val duration = exec.completionTime.getOrElse(new Date()).getTime - exec.submissionTime
     val planDetails = if (planDescription) exec.physicalPlanDescription else ""
-    val nodes = if (details) printableMetrics(graph.allNodes, exec.metricValues) else Seq.empty
+    val nodes = if (details) printableMetrics(graph.allNodes,
+      Option(exec.metricValues).getOrElse(Map.empty)) else Seq.empty
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/status/api/v1/sql/SqlResource.scala:98]: I think the format is:
```scala
    val nodes = if (details) {
      printableMetrics(graph.allNodes, Option(exec.metricValues).getOrElse(Map.empty))
    } else {
      Seq.empty
    }
```
Please see https://github.com/databricks/scala-style-guide#curly-braces:
image

**Review by wangyum — APPROVED** [review]

**Comment by wangyum** [comment]
Merged to maser, branch-3.3, branch-3.2 and branch-3.1.

**Pull request closed without merging**